### PR TITLE
Replace plethora of tools with `ruff`

### PR DESCRIPTION
**Pull request by alexdewar**
I've been looking at whether to swap our exiting tooling in `poetry_template_2` for `ruff` and thought I'd try it out on the FINESSE codebase. It found a number of small stylistic issues with the current codebase and fixed most of them automatically.

I've used it to replace:

- autoflake
- black
- isort
- flake8 (and flake8-docstrings)
- pyupgrade

In addition, I also enabled ruff's own rule set.

I was concerned that it might not be so hot with `pyupgrade`'s checks, so went back in the git history to before I enabled `pyupgrade` and ran `ruff` over the codebase. It seemed to do a pretty decent job, though there was one check that `pyupgrade` could auto-fix that `ruff` couldn't. In the grand scheme of things though, it seems nicer to just be using a single tool.

Although it found various small issues all over the place, there are no merge conflicts with other things I've been working on with FINESSE, so I thought I should get this merged.

### CODE REVIEW

**Review by jamesturner246 — APPROVED** [review]
Nice too just have one tool. No chance of them battling each other. LGTM.

File: finesse/gui/hardware_set/device_view.py [hunk 1]
```diff
@@ -319,7 +319,7 @@ def _on_open_close_clicked(self) -> None:
 class DeviceControl(QGroupBox):
     """Allows for viewing and connecting to devices."""
 
-    def __init__(self, connected_devices: AbstractSet[OpenDeviceArgs]) -> None:
+    def __init__(self, connected_devices: Set[OpenDeviceArgs]) -> None:
```
Comment (dalonsoa) [inline finesse/gui/hardware_set/device_view.py:322]: Out of curiosity, is this something you have decided to fix here simply because of convenience, or has it been a sort of recommendation/fix done by ruff? Either way, why the change?

File: finesse/hardware/device.py [hunk 2]
```diff
@@ -65,11 +65,11 @@ def get_device_types() -> dict[DeviceBaseTypeInfo, list[DeviceTypeInfo]]:
 class AbstractDevice(ABC):
     """An abstract base class for devices."""
 
-    _device_base_type_info: DeviceBaseTypeInfo
+    _device_base_type_info: ClassVar[DeviceBaseTypeInfo]
```
Comment (dalonsoa) [inline finesse/hardware/device.py:68]: What's `ClassVar` for?

**Review by dalonsoa — APPROVED** [review]
All changes look good, although I must admit I don't know the reason for some of them (the type ones, specially).

File: finesse/gui/hardware_set/device_view.py (hunk 1 above)
Comment (alexdewar) [reply finesse/gui/hardware_set/device_view.py:322]: It was a recommendation from ruff. Apparently using `typing.AbstractSet` is deprecated and we should be using `collections.abc.Set` instead.

File: finesse/hardware/device.py (hunk 2 above)
Comment (alexdewar) [reply finesse/hardware/device.py:68]: It indicates that the variable is a class variable so it should not be being changed by individual instances of `AbstractDevice`. (I just tried doing this to see what happened and `mypy` duly told me off.)

Ruff suggested this one too.
